# Getting Started with Coding Python! 

## What are variables?

* A variable is a container for storing a data value.
* Variables have a name and are associated with a value.
* Equal sign **( = )** is used to assign values to variables.

In [1]:
v = 10 

# v is the variable name
# '=' is the assignment operator
# 10 is the value

### **Variable names:**

* No spaces allowed (Use underscore or init caps, example - 'my_name' or 'myName')
* No special symbols or characters like '"@#%^&()-/\* except the underscore '_'
* Cannot begin with a number but may have a number within the variable name
* The python community by convention begin variable names with small letters but caps are allowed anywhere else

In [2]:
my_name = 'John Smith' #Using underscore in variable name

In [3]:
my name = 'John Smith' #Using space in variable name; throws an error (SyntaxError)

SyntaxError: invalid syntax (1298783299.py, line 1)

In [4]:
#Example of variables with number in variable name

v_1 = 8
v_2 = 9

To use a variable, just type out the variable name.

In [5]:
print(my_name)

John Smith


In [6]:
age = v_1 + v_2

In [7]:
print ('I am', age, 'years old.')

I am 17 years old.


**Note**: The print function is used to output data. It's a great way to see what your code is doing.

It is a function that displays its arguments (inputs). It can take multiple arguments (inputs) separated by commas ','.

We will introduce functions later.

## **Comments**

* Single line commenting symbol is the hash symbol '#'
* The interpreter ignores everything after #
* Leave comments in your code to make it understandable to other team members and for yourself

In [8]:
# This is a single-line comment

* Multi-line - Not supported in Python
* Use the # on each line of a multiline comment

In [9]:
# This is a multiline comment
# This line and the line above are both ignored by the interpreter

### 💡📝 Exercise 1: Calculate and Display Your Age in the Future. 

In [10]:
# Current year
current_year = 2024

# Your current age
current_age = 25 #Change the value in "current_age" as per your age. 

# Future year for which we want to calculate age
future_year = 2030

# Calculating future age
# This is done by subtracting the current year from the future year and adding the result to the current age
future_age = current_age + (future_year - current_year)

# Printing the results
print("Your current age is:", current_age)
print("In the year", future_year, "you will be", future_age, "years old.")


Your current age is: 25
In the year 2030 you will be 31 years old.


## Operators in Python

Python operators are used to perform operations on variables and values.

### Arithmetic Operators

In [11]:
# Addition
print("5 + 3 =", 5 + 3)

# Subtraction
print("5 - 3 =", 5 - 3)

# Multiplication
print("5 * 3 =", 5 * 3)

# Division
print("5 / 3 =", 5 / 3)

# Modulus
print("5 % 3 =", 5 % 3)

# Exponentiation
print("5 ** 3 =", 5 ** 3)

# Floor Division
print("5 // 3 =", 5 // 3)

5 + 3 = 8
5 - 3 = 2
5 * 3 = 15
5 / 3 = 1.6666666666666667
5 % 3 = 2
5 ** 3 = 125
5 // 3 = 1


### Comparison Operators

In [12]:
x = 10
y = 12

# Equal
print("x == y is", x == y)

# Not Equal
print("x != y is", x != y)

# Greater Than
print("x > y is", x > y)

# Less Than
print("x < y is", x < y)

# Greater Than or Equal to
print("x >= y is", x >= y)

# Less Than or Equal to
print("x <= y is", x <= y)

x == y is False
x != y is True
x > y is False
x < y is True
x >= y is False
x <= y is True


### Logical Operators

In [13]:
x = True
y = False

# and Operator
print("x and y is", x and y)

# or Operator
print("x or y is", x or y)

# not Operator
print("not x is", not x)

x and y is False
x or y is True
not x is False


## Data Types

### 1. Integer (`int`)

* No decimal points.
* Can be of any length (as long as it fits in memory).
* Can be used in arithmetic operations.

In [14]:
# Integer examples
x = 10
y = -3
z = 0

print("x is of type:", type(x))
print("y is of type:", type(y))
print("z is of type:", type(z))

x is of type: <class 'int'>
y is of type: <class 'int'>
z is of type: <class 'int'>


### 2. Floating Point Number (`float`)

* Contains a decimal point.
* Can represent very large or very small numbers using exponential form (e or E). For example, 1.2e34.

In [15]:
# Float examples
a = 10.5
b = -0.5
c = 1.2e34

print("a is of type:", type(a))
print("b is of type:", type(b))
print("c is of type:", type(c))

a is of type: <class 'float'>
b is of type: <class 'float'>
c is of type: <class 'float'>


In [16]:
# Converting float to integer
float_to_int = int(a)
print("Float to Integer:", float_to_int)

# Converting integer to float
int_to_float = float(x)
print("Integer to Float:", int_to_float)

Float to Integer: 10
Integer to Float: 10.0


### 3. Strings (`str`)

* In Python, strings are used to store and manipulate text data. 
* A string is a sequence of characters enclosed in quotes (either single ' or double " quotes). 

#### Creating Strings:

Strings can be created by enclosing characters inside quotes. Python treats single and double quotes the same.

In [17]:
# String examples
string_with_single_quotes = 'Hello'
string_with_double_quotes = "World"

print(string_with_single_quotes)
print(string_with_double_quotes)

Hello
World


Strings are **Immutable**. Immutability in Python means that once a string is created, it cannot be modified or changed. Any operation that appears to modify a string actually creates a new string. This concept is fundamental to understanding how strings work in Python.

In [18]:
# Define a string
original_string = "Hello, World!"

# Attempt to change a character within the string
# This will result in an error because strings are immutable
original_string[0] = 'J'  # This will raise a TypeError

TypeError: 'str' object does not support item assignment

#### Accessing and Slicing Strings:

You can access characters in a string by indexing (starting from 0) and slice strings to create substrings.

In [ ]:
example_string = "Python Programming"

# Accessing first character
first_char = example_string[0]

# Slicing from 3rd to 8th character
substring = example_string[2:8]

print("First character:", first_char)
print("Substring:", substring)

#### String Concatenation and Repetition:

Strings can be concatenated using `+` and repeated using `*`.

In [19]:
first_name = "John"
last_name = "Doe"

# Concatenation
full_name = first_name + " " + last_name

# Repetition
repeated_name = first_name * 3

print("Full Name:", full_name)
print("Repeated Name:", repeated_name)

Full Name: John Doe
Repeated Name: JohnJohnJohn


To achieve the desired result of changing a character in a string, without the "TypeError"... 

In [20]:
# Define a string
original_string = "Hello, World!"

# Create a new string with the desired modification
new_string = "J" + original_string[1:]

# Print the new string
print(new_string)  # Output: "Jello, World!"

Jello, World!


#### Common String Methods:

Python provides a variety of built-in methods for common string manipulations.

In [21]:
sample_string = "hello world"

# Capitalize
capitalized = sample_string.capitalize()

# Upper case
upper_case = sample_string.upper()

# Replace
replaced_string = sample_string.replace("world", "Python")

print("Capitalized:", capitalized)
print("Upper Case:", upper_case)
print("Replaced String:", replaced_string)

Capitalized: Hello world
Upper Case: HELLO WORLD
Replaced String: hello Python


### 💡📝 Exercise 2: Name and Hobby. 

In [22]:
# Storing name and hobby
name = "Alice"  # replace with your name or a fictional name
hobby = "reading"  # replace with your hobby

# Creating a sentence
sentence = "My name is " + name + " and I enjoy " + hobby + "."

# Printing the sentence
print(sentence)

My name is Alice and I enjoy reading.


## Control Flow 

A program’s control flow is the order in which the program’s code executes.

The control flow of a Python program is regulated by conditional statements, loops, and function calls.

Python has three types of control structures:

* Sequential - default mode
* Selection - used for decisions and branching
* Repetition - used for looping, i.e., repeating a piece of code multiple times.

### 1. Sequential: 

Sequential statements are a set of statements whose execution process happens in a sequence. The problem with sequential statements is that if the logic has broken in any one of the lines, then the complete source code execution will break.

In [23]:
## This is a Sequential statement
 
a = 20
b = 10
c = a - b

print("Subtraction is :", c)

Subtraction is : 10


### 2. Selection: 

In Python, the selection statements are also known as Decision control statements or branching statements.
The selection statement allows a program to test several conditions and execute instructions based on which condition is true.


### If statement:

It help us to run a particular code, but only when a certain condition is met or satisfied. A if only has one condition to check.

<img src="Images/Screenshot%202024-01-18%20at%207.52.54%20AM.png" alt="My Image" width="300"/>

In [24]:
x = 10
if x > 5:
    print("x is greater than 5")

x is greater than 5


### If-else statement:

The if-else statement evaluates the condition and will execute the body of if if the test condition is True, but if the condition is False, then the body of else is executed.

<img src="Images/Screenshot%202024-01-18%20at%208.01.28%20AM.png" alt="My Image" width="300"/>

In [25]:
x = 10
if x > 5:
    print("x is greater than 5")
else:
    print("x is not greater than 5")

x is greater than 5


### If-elif-else statement:

The if-elif-else chain is used for checking multiple conditions. It runs each condition until one is true or runs the else block if none are true.

<img src="Images/Screenshot%202024-01-18%20at%208.03.42 AM.png" alt="My Image" width="500"/>

In [26]:
z = 7
if z > 10:
    print("z is greater than 10")
elif z < 5:
    print("z is less than 5")
else:
    print("z is between 5 and 10")

z is between 5 and 10


### 3. Repetition

A repetition statement is used to repeat a group (block) of programming instructions.

### for loop

A way to repeat a block of code a specific number of times. 

<img src="Images/Screenshot%202024-01-18%20at%208.12.33%20AM.png" alt="My Image" width="300"/>

In [27]:
for i in range(5):
    print(i)

0
1
2
3
4


`for`: Tells us we’re going to loop through one by one

`i`: A variable that helps us keep track of where we are (index)

`(5)`: Number of iterations; can be a variable, as long as it’s an int!

`range`: Built-in function
 * iterates through 0,1,2,3,4
 * assumes 0 is the start index
 * end_index is not inclusive!

### while loop

while loops repeatedly execute a block of code as long as a condition is true.

<img src="Images/Screenshot%202024-01-18%20at%208.14.10%20AM.png" alt="My Image" width="300"/>

In [28]:
count = 0
while count < 3:
    print(count)
    count = count + 1

0
1
2


### 💡📝 Exercise 3: Write a program that prints all integers between 100 and 125 (including 100 and 125) to the screen using a for loop. Also, can you print even numbers from 100 to 125? [Hint: range function takes third argument that specifies the step (increments by 2 for even numbers)]

### 💡📝 Exercise 4: Do the same as exercise 3 (not the even number part), but this time use a while loop.

A similar solution is provided below to print numbers from 1 to 10. 

In [29]:
for i in range(1, 11):
    print (i)

1
2
3
4
5
6
7
8
9
10


In [30]:
i = 1
while i < 11:
    print (i)
    i = i + 1

1
2
3
4
5
6
7
8
9
10


## Data Structures in Python

In Python, data structures are used to store and organize data in a way that facilitates efficient access and modification. There are several built-in data structures in Python, each with its own characteristics and use-cases.

### 1. List

* Description: A list is a collection that is ordered and changeable. It allows duplicate members.
* Common Uses: Lists are used for storing a sequence of elements that need to be accessed by their position or index.
* Creating lists - Put comma separated values within square brackets. Values can be of different data types.

In [31]:
list1 = ['physics', 'astronomy', 56.98, 'MJ', -9.36]

# list within a list
list2 = ['chemistry', 'biology', 77.98, [3, 4, 65], 'EE', -56]

#### Accessing values in a list via indexing: 
    
* Same as String objects.
* Elements or a range of elements can be accessed by index.

In [32]:
list1[0]

'physics'

In [33]:
list1[1]

'astronomy'

In [34]:
list1[4]

-9.36

In [35]:
list1[-1]

-9.36

In [36]:
list1[1][3]

'r'

In [37]:
list2[3]

[3, 4, 65]

In [38]:
list2[3][2]

65

In [39]:
list1[0:4]

['physics', 'astronomy', 56.98, 'MJ']

In [40]:
list2[3:5]

[[3, 4, 65], 'EE']

In [41]:
[1,2,3] + [5,6,7] #List Concatenation 

[1, 2, 3, 5, 6, 7]

In [42]:
list1 + list2

['physics',
 'astronomy',
 56.98,
 'MJ',
 -9.36,
 'chemistry',
 'biology',
 77.98,
 [3, 4, 65],
 'EE',
 -56]

#### Some Lists Methods

In [43]:
#Length of the list
len(list1)

5

In [44]:
#Maximum/Minimum value
list3 = [3, 4, 67, 2, 1]
list4 = ['3', '4', '67', '2', '1', 'four', 'three']

max(list3)

67

In [45]:
min(list3)

1

In [46]:
#Append object to list
list4.append('two')
list4.append('three')
list4

['3', '4', '67', '2', '1', 'four', 'three', 'two', 'three']

In [47]:
#Insert object to list
list4.insert(3, '56') #(index,value)
list4

['3', '4', '67', '56', '2', '1', 'four', 'three', 'two', 'three']

In [48]:
#Frequency of object
list4.count('three')

2

In [49]:
#Return index
list4.index('1')

5

In [50]:
#Delete object: by index position
list3

[3, 4, 67, 2, 1]

In [51]:
list3.pop(3)

2

In [52]:
list3

[3, 4, 67, 1]

In [53]:
#Delete object: by element referencing
list3.remove(67)

In [54]:
list3

[3, 4, 1]

In [55]:
#Reverse a list
list3.reverse()
list3

[1, 4, 3]

In [56]:
#Sort the list
list3.sort()
list3

[1, 3, 4]

In [57]:
#Lists are "MUTABLE"
list3[2] = 5
list3

[1, 3, 5]

#### List Comprehension

When programming, frequently we want to transform one type of data into another. As a simple example, consider the following code that computes square numbers:

In [58]:
nums = [0, 1, 2, 3, 4]
squares = []
for x in nums:
    squares.append(x ** 2)
print(squares)

[0, 1, 4, 9, 16]


In [59]:
nums = [0, 1, 2, 3, 4]
squares = [x ** 2 for x in nums]
print(squares)

[0, 1, 4, 9, 16]


In [60]:
#List comprehensions can also contain conditions:

nums = [0, 1, 2, 3, 4]
even_squares = [x ** 2 for x in nums if x % 2 == 0]
print(even_squares)

[0, 4, 16]


### 2. Tuples

* Description: A tuple is a collection that is ordered and unchangeable. It also allows duplicate members.
* Common Uses: Tuples are used when you need an ordered collection that cannot be changed, such as a set of constant values.

* Thus, tuples are sequences, just like lists except: 

    - Tuples are immutable - cannot be changed or updated like lists.
    - Tuples use parentheses ( ), whereas lists use square brackets [ ].

In [61]:
tup = (3, 4, 67, 2, 1)

In [62]:
tup[2] = 8

TypeError: 'tuple' object does not support item assignment

### 3. Sets

* Description: A set is a collection that is unordered, unindexed, and unchangeable, but new items can be added. Sets do not allow duplicate elements.
* Common Uses: Sets are used for membership testing, removing duplicates from a sequence, and computing mathematical operations like intersection, union, difference, and symmetric difference.
* Sets are defined using '{ }'.

In [63]:
animals = {'cat', 'dog'}
print('cat' in animals)   # Check if an element is in a set; prints "True"
print('fish' in animals)  # prints "False"

True
False


In [64]:
animals.add('fish')      # Add an element to a set
print('fish' in animals)
print(len(animals))       # Number of elements in a set;

True
3


In [65]:
animals.add('cat')       # Adding an element that is already in the set does nothing
print(len(animals))       
animals.remove('cat')    # Remove an element from a set
print(len(animals))   

3
2


Iterating over a set has the same syntax as iterating over a list; however since sets are unordered, you cannot make assumptions about the order in which you visit the elements of the set:

In [66]:
animals = {'cat', 'dog', 'fish'}
for idx, animal in enumerate(animals):
    print('#{}: {}'.format(idx + 1, animal))

#1: dog
#2: cat
#3: fish


Like lists, we can easily construct sets using set comprehensions:

In [67]:
from math import sqrt
print({int(sqrt(x)) for x in range(30)})

{0, 1, 2, 3, 4, 5}


### 4. Dictionary

* Description: A dictionary is a collection that is ordered (as of Python 3.7) and changeable. It stores data in key-value pairs.
* Common Uses: Dictionaries are used when you need a logical association between a key:value pair. They are optimized for retrieving data when you know the key.
* Similar to a list except values are accessed by looking up a **key** instead of an index.
* A key can be a string or number.
* Creating dictionaries - **key-value** pairs are separated by colons (**:**), items are separated by commas (**,**) and everything is enclosed in curly braces { }.

In [68]:
d = {'cat': 'cute', 'dog': 'furry'}  # Create a new dictionary with some data
print(d['cat'])       # Get an entry from a dictionary; prints "cute"
print('cat' in d)     # Check if a dictionary has a given key; prints "True"

cute
True


In [69]:
d2 = {'name':'Francis','age':60,'job':'Car designer','brand':'Jaguar','worked-for':['Ford','TWR','Aston Martin']}

In [70]:
d2['worked-for']

['Ford', 'TWR', 'Aston Martin']

In [71]:
d2['worked-for'][1]

'TWR'

In [72]:
#Dictionaries are MUTABLE.

d['fish'] = 'wet'    # Set an entry in a dictionary
print(d)
print(d['fish'])      # Prints "wet"

{'cat': 'cute', 'dog': 'furry', 'fish': 'wet'}
wet


In [73]:
d['fish'] = 'water'    # Modify an entry 
print(d)
print(d['fish'])      # Prints "water"

{'cat': 'cute', 'dog': 'furry', 'fish': 'water'}
water


In [74]:
print(d['monkey'])  # KeyError: 'monkey' not a key of d

KeyError: 'monkey'

#### Some dictionary methods

In [75]:
#Length of a dictionary
dict1 = {"name":"Daniel","age":23,"degree":"MS"}
dict2 = {'name':'Francis','age':60,'job':'Car designer','brand':'Jaguar','worked-for':['Ford','TWR','Aston Martin']}

len(dict1)

3

In [76]:
#Return value with get method

dict1.get('name')

'Daniel'

In [77]:
#List of key-value pairs

dict1.items()

dict_items([('name', 'Daniel'), ('age', 23), ('degree', 'MS')])

In [78]:
#List of keys

dict1.keys()

dict_keys(['name', 'age', 'degree'])

In [79]:
#List of values

dict1.values()

dict_values(['Daniel', 23, 'MS'])

Dictionary comprehensions: These are similar to list comprehensions, but allow you to easily construct dictionaries. For example:



In [80]:
nums = [0, 1, 2, 3, 4]
even_num_to_square = {x: x ** 2 for x in nums if x % 2 == 0}
print(even_num_to_square)

{0: 0, 2: 4, 4: 16}


### 💡📝 Exercise 5: Print the largest element in a given list. 


In [81]:
#Here's a helper code to get the smallest element in the list.

list_numbers = [67, 42, 53, 74, 55]
list_numbers.sort()
list_numbers[0]

42

### 💡📝 Exercise 6: Fruit Count Dictionary.

You have a list of fruits, and you want to create a dictionary to store the count of each type of fruit in the list. Write Python code to create a dictionary named `fruit_counts` that counts the occurrences of each fruit in the list.

fruits = ['apple', 'banana', 'orange', 'apple', 'banana', 'apple', 'pear']

In [82]:
#output: {'apple': 3, 'banana': 2, 'orange': 1, 'pear': 1}

In [83]:
# Step 1: Initialize an empty dictionary
fruit_counts = {}

# Step 2: Create a list of fruits
fruits = ['apple', 'banana', 'orange', 'apple', 'banana', 'apple', 'pear']

# Step 3: Iterate through the list of fruits
for fruit in fruits:
    # Step 4: Check if the fruit is in the dictionary and update the count
    if fruit in fruit_counts:
        fruit_counts[fruit] += 1
    else:
        # If the fruit is not in the dictionary, add it with an initial count of 1
        fruit_counts[fruit] = 1

# Step 5: Print the fruit_counts dictionary
print(fruit_counts)

{'apple': 3, 'banana': 2, 'orange': 1, 'pear': 1}


## Functions in Python

* In Python, a function is a block of code that performs a specific task or set of tasks. 
* Functions allow you to write reusable code, making your programs more organized and efficient.
* Functions in Python are defined using the def keyword, followed by the function name and parentheses which can contain parameters. 

### Function Analogy 

<img src="Images/Screenshot%202024-01-29%20at%209.11.48%20PM.png" alt="My Image" width="400"/>

<img src="Images/Screenshot%202024-01-29%20at%209.11.54%20PM.png" alt="My Image" width="500"/>

<img src="Images/Screenshot%202024-01-29%20at%209.12.00%20PM.png" alt="My Image" width="700"/>

#### Function definition:

In [84]:
# def function_name(param1, param2):
#     result = do something
#     return result

* `function_name`: name
* `param1, param2`: parameters (input expected): One or more variables that a function expects as input.
* `result`: output/return value

<img src="Images/Screenshot%202024-01-29%20at%209.16.00%20PM.png" alt="My Image" width="700"/>

### 💡📝 Exercise 7: Find the function definition, function name, parameter(s), and return value in average.

In [85]:
def main():
    mid = average(10.6, 7.2)
    print(mid)

def average(a, b):
    sum = a + b
    return sum / 2

In [86]:
# Function definition:

# Name: 

# Parameters: 

# Return value: 

### Anatomy of a function:

* Return value: Value that a function hands back to the “calling” function. 
* main() is the caller (calling function); average is the callee (called function).
* mid = average(10.6, 7.2) is the function call. 
* 10.6, 7.2 are the arguments. 

* What’s the difference between arguments and parameters?
    - parameters are the name of input values in the function definition.
    - arguments are the values passed in when function is called!

### How to use Functions: 

In [87]:
def my_function():
    # code goes here
    print("Hello from a function")

In [88]:
my_function()  # Output: Hello from a function

Hello from a function


In [89]:
def add_numbers(a, b):
    return a + b

result = add_numbers(3, 4)
print(result)  # Output: 7

7


### Variable Scope

Variable scope refers to the areas in your code where a variable can be accessed. There are two main scopes:
There are two main scopes:

1. Local Scope: The variable is only accessible within the function where it is defined.
2. Global Scope: The variable is accessible anywhere in the program.

In [90]:
def my_function():
    local_variable = 5
    print(local_variable)  # Output: 5

my_function()
# print(local_variable)  # This would cause an error because local_variable is not accessible outside the function.

5


In [91]:
global_variable = 10

def my_function():
    print(global_variable)  # Output: 10

my_function()
print(global_variable)  # Output: 10

10
10


### 💡📝 Exercise 8: Temperature Converter

Create a function named `convert_temperature` that converts a temperature from Fahrenheit to Celsius.

In [92]:
# def convert_temperature(fahrenheit):
#     # Hint: Use the formula Celsius = (Fahrenheit - 32) * 5/9 to convert the temperature.
#     # Make sure to return the result.

# # Example test
# fahrenheit_temperature = 68
# celsius_temperature = convert_temperature(fahrenheit_temperature)
# celsius_temperature

### 💡📝 Exercise 9: Average Calculator

1. Create a Function: Write a function named calculate_average that accepts a list of numbers and returns the average.
2. Local Variable: Inside the function, create a local variable sum to hold the sum of the numbers.
3. Global Variable: Define a global variable data which is a list of numbers.
4. Call the Function: Call calculate_average using data and print the result.
5. Bonus: Handle an empty list gracefully by returning None.

In [93]:
# Defining the global variable 'data'
data = [1, 2, 3, 4, 5]

def calculate_average(numbers):
    if not numbers:  # Check if the list is empty
        return None
    sum = 0  # Initializing the local variable 'sum'
    for number in numbers:
        sum += number  # Adding each number to the sum
    average = sum / len(numbers)  # Calculating the average
    return average

# Calling the function and printing the result
average_result = calculate_average(data)
average_result

3.0

## File Handling in Python

File handling in Python is a fundamental aspect for many applications, from simple data processing to complex file manipulation. 
Python provides built-in functions for reading, writing, and manipulating files.

* Opening a File: The open() function is used to open a file. It requires the file path and the mode of operation as arguments.

* Modes of Operation:

  - 'r' - Read mode (default mode; opens a file for reading).
  - 'w' - Write mode (opens a file for writing, creates a new file if it doesn't exist, or truncates the file if it does).
  - 'a' - Append mode (opens a file for appending at the end; creates a new file if it doesn't exist).
  - 'b' - Binary mode (used for binary files, e.g., 'rb' or 'wb').
  - '+' - Update mode (for both reading and writing).

* Closing a File: It's crucial to close a file after your operations to free up system resources. Use the close() method.

* The with Statement: This is a better practice for opening files as it automatically takes care of closing the file after the block of code is executed, even if an exception occurs.

### Reading from a File

In [94]:
# Using the with statement for safe file handling
with open('example.txt', 'r') as file:
    content = file.read()
    print(content)

Hello, World!


* read() reads the entire file.
* readline() reads the next line from the file.
* readlines() reads all the lines into a list.

### Writing to a File

In [95]:
with open('example.txt', 'w') as file:
    file.write("Hello, World!")

* write() writes a string to the file.
* For writing multiple lines, you can use writelines() or multiple write() calls.
* Use 'rb' or 'wb' mode for reading or writing binary files, such as images or documents.

In [96]:
# First, let's create a sample text file to work with
sample_text = """Line 1
Line 2
Line 3
Line 4"""

# Writing the sample text to 'sample_file.txt'
with open('sample_file.txt', 'w') as file:
    file.write(sample_text)

# Now, let's write the script to count the lines and append the count to the file
file_path = 'sample_file.txt'

# Reading the file and counting the lines
with open(file_path, 'r') as file:
    lines = file.readlines()
    line_count = len(lines)

# Appending the line count to the file
with open(file_path, 'a') as file:
    file.write(f"\nTotal number of lines: {line_count}")

# For demonstration, let's read the file again to show the appended text
with open(file_path, 'r') as file:
    updated_content = file.read()

updated_content

'Line 1\nLine 2\nLine 3\nLine 4\nTotal number of lines: 4'

### References

1. Severance, C. "Programming for Everybody (Getting Started with Python)." University of Michigan. Available at: https://www.coursera.org/learn/python

2. Traversy, B. 2018. "Python Crash Course for Beginners." YouTube. Available at: https://www.youtube.com/watch?v=JJmcL1N2KQs

3. Sweigart, A. "Automate the Boring Stuff with Python." Available at: https://automatetheboringstuff.com/